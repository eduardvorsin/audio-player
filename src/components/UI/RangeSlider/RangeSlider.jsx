import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import { StyledRangeSlider } from "./StyledRangeSlider";

export const RangeSlider = forwardRef(({
  className,
  onChange,
  isVertical,
  testId,
  ...props
}, ref) => {

  const changeHandler = (e) => {
    const calculatedPercent =
      e.target.max === '' ? e.target.value : e.target.value * 100 / e.target.max;

    e.target.style = `--progress-percent:${calculatedPercent}%;`;

    onChange(e);
  }

  const currentPercent = props.value || 0;

  return (
    <StyledRangeSlider
      className={className}
      isVertical={isVertical}
      data-testid={testId}
    >
      <input
        style={{ '--progress-percent': `${currentPercent}%` }}
        ref={ref}
        type='range'
        onChange={changeHandler}
        {...props}
      />
    </StyledRangeSlider>
  );
});

RangeSlider.propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func,
  isVertical: PropTypes.bool,
  testId: PropTypes.string,
};

RangeSlider.defaultProps = {
  className: '',
  onChange: () => { },
  isVertical: false,
  testId: ''
};
