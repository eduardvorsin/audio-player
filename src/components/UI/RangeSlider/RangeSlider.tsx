import React, { forwardRef } from "react";
import { StyledRangeSlider } from "./StyledRangeSlider";

interface RangeSliderProps extends Partial<React.InputHTMLAttributes<HTMLInputElement>> {
  className?: string,
  onChange?: React.ChangeEventHandler<HTMLInputElement>,
  defaultValue?: number;
  isVertical?: boolean,
}

interface CSSPropertiesWithVars extends React.CSSProperties {
  '--progress-percent': string,
}

export type RangeSliderRef = HTMLInputElement;

export const RangeSlider = forwardRef<RangeSliderRef, RangeSliderProps>(({
  className = '',
  onChange = () => { },
  isVertical = false,
  defaultValue = 0,
  ...props
}, ref) => {

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const valueAsNumber = Number(e.target.max);
    const maxAsNumber = Number(e.target.max);

    const calculatedPercent: number =
      maxAsNumber === 0 ? valueAsNumber : valueAsNumber * 100 / maxAsNumber;

    e.target.setAttribute('style', `--progress-percent:${calculatedPercent}%;`);

    onChange(e);
  }

  const sliderStyles: CSSPropertiesWithVars =
    { '--progress-percent': `${defaultValue}%` };

  return (
    <StyledRangeSlider
      className={className}
      isVertical={isVertical}
      data-testid='range-slider'
    >
      <input
        style={sliderStyles}
        ref={ref}
        defaultValue={defaultValue}
        type='range'
        onChange={changeHandler}
        {...props}
      />
    </StyledRangeSlider>
  );
});