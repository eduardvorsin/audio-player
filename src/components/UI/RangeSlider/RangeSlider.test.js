import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { RangeSlider } from "./RangeSlider";

describe('Range slider component tests', () => {

  test('renders correctly', () => {
    render(<RangeSlider />);
    expect(screen.getByRole('slider')).toBeInTheDocument();
  });

  test('at the onChange event, the mock function is triggered', () => {
    const mockFn = jest.fn();
    render(
      <RangeSlider
        onChange={mockFn}
      />
    );

    fireEvent.change(screen.getByRole('slider'), { target: { value: 50 } })
    expect(screen.getByRole('slider')).toHaveValue('50');
  });

  test('if the isVertical prop is passed', () => {
    render(
      <RangeSlider isVertical />
    );
    expect(screen.getByRole('slider')).toHaveStyle('writing-mode: bt-lr');
  });

  test('className is assigned correctly', () => {
    render(
      <RangeSlider
        className='test'
      />
    );
    expect(screen.getByTestId('range-slider')).toHaveClass('test');
  });

  test('snapshot with default prop values', () => {
    render(
      <RangeSlider />
    );

    expect(screen.getByRole('slider')).toMatchSnapshot();
  });

  test('snapshot with vertical slider', () => {
    render(
      <RangeSlider isVertical />
    );

    expect(screen.getByRole('slider')).toMatchSnapshot();
  });
});
