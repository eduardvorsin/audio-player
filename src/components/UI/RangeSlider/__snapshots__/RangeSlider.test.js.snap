// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range slider component tests snapshot with default prop values 1`] = `
<input
  style="--progress-percent: 0%;"
  type="range"
/>
`;

exports[`Range slider component tests snapshot with vertical slider 1`] = `
<input
  style="--progress-percent: 0%;"
  type="range"
/>
`;
